Uses an auxiliary script for token insertion.

jmchamica opened the pull request:
In the previous implementation, the JSON claims object was stored as a
string value and later delivered to the `docker exec ... psql` through
the shell.
Since Windows and Unix shells handle escaping of characters
differently, a Unix shell script was written for use inside the
container. This way, the JSON string is created inside the container.

EDIT:
After all the issues with incompatible shells, file and text formats (\r\n vs \n) and so on, the better solution for executing the token insertion tasks is (imo) using JDBC. This solves the multiplatform issue, as well as the JDBC_DATABASE_URL caching issue.  

closes #183

jmchamica committed:
Uses an auxiliary script for token insertion.

In the previous implementation, the JSON claims object was stored as a
string value and later delivered to the `docker exec ... psql` through
the shell.
Since Windows and Unix shells handle escaping of characters
differently, a Unix shell script was written for use inside the
container. This way, the JSON string is created inside the container.

JorgePBrown commented:
This isn't working on Windows.
```
> Task :pgInsertReadToken FAILED
string token:MLMb1TI4aNw9me9ahjjgBuPueSQwK89YBZdUBjv3M4Q
hash:ce5cdd75a0dc5129806c5e4bfa956c8d933f9bde6b78b0437e3bb83a7cb195e4
/usr/local/bin/docker-entrypoint.sh: /mnt/sh/insert_token: /bin/sh^M: bad interpreter: No such file or directory
/usr/local/bin/docker-entrypoint.sh: line 315: /mnt/sh/insert_token: Success
```

Not sure what error this is. The file system in the docker container should always be the same.

jmchamica commented:
> This isn't working on Windows.
> 
> ```
> > Task :pgInsertReadToken FAILED
> string token:MLMb1TI4aNw9me9ahjjgBuPueSQwK89YBZdUBjv3M4Q
> hash:ce5cdd75a0dc5129806c5e4bfa956c8d933f9bde6b78b0437e3bb83a7cb195e4
> /usr/local/bin/docker-entrypoint.sh: /mnt/sh/insert_token: /bin/sh^M: bad interpreter: No such file or directory
> /usr/local/bin/docker-entrypoint.sh: line 315: /mnt/sh/insert_token: Success
> ```
> 
> Not sure what error this is. The file system in the docker container should always be the same.

At glance, seems like a \r\n vs \n problem with the shell script.
Anyway, thanks for testing! I'll probably set a VM and try again.

JorgePBrown commented:
> > This isn't working on Windows.
> > ```
> > > Task :pgInsertReadToken FAILED
> > string token:MLMb1TI4aNw9me9ahjjgBuPueSQwK89YBZdUBjv3M4Q
> > hash:ce5cdd75a0dc5129806c5e4bfa956c8d933f9bde6b78b0437e3bb83a7cb195e4
> > /usr/local/bin/docker-entrypoint.sh: /mnt/sh/insert_token: /bin/sh^M: bad interpreter: No such file or directory
> > /usr/local/bin/docker-entrypoint.sh: line 315: /mnt/sh/insert_token: Success
> > ```
> > 
> > 
> > Not sure what error this is. The file system in the docker container should always be the same.
> 
> At glance, seems like a \r\n vs \n problem with the shell script.
> Anyway, thanks for testing! I'll probably set a VM and try again.

The docker host in Windows runs on a Linux VM supported by Hyper-V so this shouldn't be a problem.

How does the postgres Docker image have some "extra" commands, like when running `docker exec createdb`? Maybe we can add a command like that to insert a token to our image. I assume they do it by having a script file called `createdb` in the container's PATH environment variable.

jmchamica commented:
> > > This isn't working on Windows.
> > > ```
> > > > Task :pgInsertReadToken FAILED
> > > string token:MLMb1TI4aNw9me9ahjjgBuPueSQwK89YBZdUBjv3M4Q
> > > hash:ce5cdd75a0dc5129806c5e4bfa956c8d933f9bde6b78b0437e3bb83a7cb195e4
> > > /usr/local/bin/docker-entrypoint.sh: /mnt/sh/insert_token: /bin/sh^M: bad interpreter: No such file or directory
> > > /usr/local/bin/docker-entrypoint.sh: line 315: /mnt/sh/insert_token: Success
> > > ```
> > > 
> > > 
> > > Not sure what error this is. The file system in the docker container should always be the same.
> > 
> > 
> > At glance, seems like a \r\n vs \n problem with the shell script.
> > Anyway, thanks for testing! I'll probably set a VM and try again.
> 
> The docker host in Windows runs on a Linux VM supported by Hyper-V so this shouldn't be a problem.
> 
> How does the postgres Docker image have some "extra" commands, like when running `docker exec createdb`? Maybe we can add a command like that to insert a token to our image. I assume they do it by having a script file called `createdb` in the container's PATH environment variable.

The problem is that the container (which is running against a Linux kernel) has to run a shell script accessed from the directory on your Windows host. Text files in windows end lines with \r\n as opposed to its UNIX variants (\n). The shell from within the container cannot interpret the script because it has this format.

Using COPY from a dockerfile will make the proper conversions, but in this case, we are mounting the directory from your Win host to the Linux container, so, no  conversions are made.

jmchamica committed:
Use dedicated docker image for Token Insertion.

Previously, we were mounting a directory from the host to the container
and running a script from that directory.
However, the differences between Win vs UNIX in file formatting
(specifically \r\n vs \n) would end up in error if the container
(running against a Linux kernel) attempted to run the script located
in the Win host.

JorgePBrown commented:
Seems to not have worked unfortunately.

`standard_init_linux.go:211: exec user process caused "no such file or directory"`

jmchamica commented:
> Seems to not have worked unfortunately.
> 
> `standard_init_linux.go:211: exec user process caused "no such file or directory"`

Yickes. [The problem persists event after COPYing](https://stackoverflow.com/questions/51508150/standard-init-linux-go190-exec-user-process-caused-no-such-file-or-directory).

jmchamica committed:
Use JDBC instead of Docker+Scripts.

Solves:
  - multiplatform (incompatible Shells, text and file formats of
  different OSs were conflicting with the gradle tasks)
  - recheck the JDBC_DATABASE_URL every time, instead of caching
  configurations

JorgePBrown commented:
Ladies and gentlemen, we got them.

```
> Task :pgInsertReadToken
string token:zkjk2KhohdhlcB83QfiCUwdTO5EP-v8UzqTMiM1T-EU
hash:828f253efc7d12591a697a255cee4d80e11dff96b3404c0ee69e94acac8d41c2
Your token reference is zkjk2KhohdhlcB83QfiCUwdTO5EP-v8UzqTMiM1T-EU
BUILD SUCCESSFUL in 761ms
1 actionable task: 1 executed
5:36:19 PM: Task execution finished 'pgInsertReadToken'.
```

JorgePBrown commented on project/src/test/resources/docker/sh/insert_token:
```diff
@@ -0,0 +1,23 @@
+#!/bin/sh
+
+set -e
+
+nargs=6
+if [ "$#" -ne "$nargs" ]; then
+  printf "%s arguments are required:\n\t./%s <hash> <is_valid> <issued_at> <expires_at> <scope> <client_id>\n" \
+    "$nargs" "$(basename "$0")"
+  exit 1
+fi
+
+if [ -z "$CORE_DB_HOST" ] || [ -z "$CORE_DB_USER" ] || [ -z "$CORE_DB_NAME" ] || [ -z "$CORE_DB_PORT" ]; then
+  printf "Missing environment variables for DB connectivity:\n\tCORE_DB_HOST, CORE_DB_USER, CORE_DB_NAME or CORE_DB_PORT\n"
+  exit 1
+fi
+
+psql -w -1 \
+  -h "$CORE_DB_HOST" \
+  -U "$CORE_DB_USER" \
+  -d "$CORE_DB_NAME" \
+  -p "$CORE_DB_PORT" \
+  -c "INSERT INTO dbo.Token(hash,isValid,issuedAt,expiresAt,claims) VALUES ('$1','$2',$3,$4,'{\"scope\":\"$5\", \"client_id\": $6}')"
+
```
This file is useless now right?

jmchamica commented:
> Ladies and gentlemen, we got them.
> 
> ```
> > Task :pgInsertReadToken
> string token:zkjk2KhohdhlcB83QfiCUwdTO5EP-v8UzqTMiM1T-EU
> hash:828f253efc7d12591a697a255cee4d80e11dff96b3404c0ee69e94acac8d41c2
> Your token reference is zkjk2KhohdhlcB83QfiCUwdTO5EP-v8UzqTMiM1T-EU
> BUILD SUCCESSFUL in 761ms
> 1 actionable task: 1 executed
> 5:36:19 PM: Task execution finished 'pgInsertReadToken'.
> ```

ᕦ( ͡͡~͜ʖ ͡° )ᕤ

JorgePBrown commented on project/buildSrc/src/main/kotlin/db.kt:
```diff
@@ -265,103 +269,107 @@ open class PgAddData : AbstractTask() {
-
-        return listOf(insertQuery, base64Token)
+    fun create(scope: String) {
+        val params = getTokenReferences(scope)
+        val cid = 500
+        val json = PGobject(); json.type = "jsonb"; json.value = """{"scope":"$scope", "client_id": ${cid}}"""
+        val con = Db.getConnection()
+        val st = con.prepareStatement("INSERT INTO dbo.Token(hash,isValid,issuedAt,expiresAt,claims) VALUES (?,?,?,?,?);")
+        st.setString(1, params.hash)
+        st.setBoolean(2, params.isValid)
+        st.setLong(3, params.issuedAt)
+        st.setLong(4, params.expiredAt)
+        st.setObject(5, json)
+        print("Your token reference is ${params.base64Token}")
```
The SQL statement is not being executed and the connection is not being closed.

jmchamica replied:
Yickes. Completely forgot. It should execute though?

JorgePBrown replied:
🤔 If we want it to do something I think it should.

jmchamica replied:
Haha. Turns out, I removed the line that executed the statement just as I committed the changes. *brain dead*

jmchamica committed:
Cleanup and fixing a mistake.
